src/lib/tools.ts

```diff
@@ -0,0 +1,313 @@
+	NotImplementedError,
+	UnsupportedError,
+	URILoadError,
+} from './error-types';
+
+import { TimeoutError } from 'bluebird';
+import * as yamljs from 'yamljs';
+import { ConfigMap } from './controllers/config-store/config-store';
+
+export const tunnelAsync = Bluebird.promisify(tunnel);
+
+/**
+ * Returns an absolute path for a path, when in basePath
+ * @param {string} path: The file path, (or an absolute path)
+ * @param {string} basePath: The base path
+ * @returns {string} An absolute path
+ */
+export function getAbsolutePath(path: string, basePath: string): string {
+	return isAbsolute(path) ? path : join(basePath, path);
+}
+
+/**
+ * Gets absolute URI
+ * @param {string} uri
+ * @param {string} basePath
+ * @returns {string}
+ */
+export function getAbsoluteUri(uri: string, basePath: string): string {
+	if (isValidGitUri(uri)) {
+		throw new NotImplementedError('Git URI support not implemented yet');
+	} else if (isLocalPathUri(uri)) {
+		return getAbsolutePath(uri, basePath);
+	} else {
+		throw new UnsupportedError('URI type not supported yet');
+	}
+}
+
+/**
+ * Loads a yaml file as object
+ * @param {string} filePath
+ * @param {string} errorMessage
+ * @returns {Promise<string>}
+ */
+export async function loadFromFile(
+	filePath: string,
+	errorMessage: string = '',
+): Promise<object> {
+	try {
+		return yamljs.parse(await fs.readFile(filePath, 'utf8'));
+	} catch (e) {
+		throw new FileLoadError(errorMessage + e.message);
+	}
+}
+
+/**
+ * Checks uri is a valid git URI
+ * @param {string} uri
+ * @returns {boolean}
+ */
+export function isValidGitUri(uri: string): boolean {
+	return /(?:git|ssh|http|https|git@[\w\.]+):(?:\/\/)?[\w\.@\:/\-]+/.test(uri);
+}
+
+/**
+ * Checks if uri is a valid local path URI
+ * @param {string} uri
+ * @returns {boolean}
+ */
+export function isLocalPathUri(uri: string): boolean {
+	return /^[a-zA-Z0-9_/\-.]+$/.test(uri);
+}
+
+/**
+ * Loads a file in path of URI
+ * @param {string} uri
+ * @param {string} path
+ * @param {string} errorMessage
+ * @returns {Promise<string>}
+ */
+export async function loadFromUri({
+	uri,
+	path,
+	errorMessage,
+}: {
+	uri: string;
+	path: string;
+	errorMessage?: string;
+}): Promise<object> {
+	// TODO: support git URI
+	if (isValidGitUri(uri)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	}
+	if (isLocalPathUri(uri)) {
+		return await loadFromFile(join(uri, path), errorMessage);
+	}
+	throw new URILoadError(`Error loading ${path} from ${uri}`);
+}
+
+/**
+ * Gets base path of a path
+ * @param {string} path
+ * @returns {string}
+ */
+export function getBasePath(path: string): string {
+	return dirname(resolve(path));
+}
+
+/**
+ * Converts relative paths of any object to absolute paths
+ * @param {object} conf
+ * @param {string} basePath
+ * @returns {object}
+ */
+export function convertRelativePaths({
+	conf,
+	basePath,
+}: {
+	conf: any;
+	basePath: string;
+}): any {
+	// Convert relative to absolute URIs
+	const keys = [
```

Comment:
I think it'd be good to type this like
```suggestion
	const keys: Array<_.PropertyPath> = [
```
to make sure they're valid lodash property paths